File: python/min_stack.py
class MinStack:

    def __init__(self):
        self._stack: list[int] = []
        self._min_stack: list[int] = []

    def push(self, val: int) -> None:
        """time complexity: O(1), space complexity: O(n)"""
        self._stack.append(val)
        if len(self._min_stack) > 0 and val > self._min_stack[-1]:
            self._min_stack.append(self._min_stack[-1])
        else:
            self._min_stack.append(val)

    def pop(self) -> None:
        """time complexity: O(1), space complexity: O(n)"""
        self._stack.pop()
        self._min_stack.pop()

    def top(self) -> int:
        """time complexity: O(1)"""
        return self._stack[-1]

    def getMin(self) -> int:
        """time complexity: O(1)"""
        return self._min_stack[-1]


def test_min_stack_case_1():
    # arrange
    obj = MinStack()
    expected1 = -3
    expected2 = 0
    expected3 = -2

    # act
    obj.push(-2)
    obj.push(0)
    obj.push(-3)
    actual1: int = obj.getMin()
    obj.pop()
    actual2: int = obj.top()
    actual3: int = obj.getMin()

    # assert
    assert actual1 == expected1
    assert actual2 == expected2
    assert actual3 == expected3


def test_min_stack_case_2():
    # arrange
    obj = MinStack()
    expected1 = -2
    expected2 = -1
    expected3 = -2

    # act
    obj.push(-2)
    obj.push(0)
    obj.push(-1)
    actual1: int = obj.getMin()
    actual2: int = obj.top()
    obj.pop()
    actual3: int = obj.getMin()

    # assert
    assert actual1 == expected1
    assert actual2 == expected2
    assert actual3 == expected3


def test_min_stack_case_3():
    # arrange
    obj = MinStack()
    expected1 = -3
    expected2 = -3

    # act
    obj.push(-2)
    obj.push(-2)
    obj.push(-3)
    obj.push(-3)
    # [-2, -2, -3, -3]
    actual1: int = obj.getMin()
    obj.pop()
    actual2: int = obj.getMin()

    # assert
    assert actual1 == expected1
    assert actual2 == expected2

File: rust/src/code/min_stack.rs
struct MinStack {
    stack: Vec<i32>,
    min_stack: Vec<i32>,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MinStack {
    fn new() -> Self {
        Self {
            stack: Vec::<i32>::new(),
            min_stack: Vec::<i32>::new(),
        }
    }

    fn push(&mut self, val: i32) {
        self.stack.push(val);

        if self.min_stack.is_empty() {
            self.min_stack.push(val);
        } else if val <= *self.min_stack.last().unwrap() {
            self.min_stack.push(val);
        } else {
            self.min_stack.push(*self.min_stack.last().unwrap());
        }
    }

    fn pop(&mut self) {
        self.stack.pop().unwrap();
        self.min_stack.pop().unwrap();
    }

    fn top(&self) -> i32 {
        *self.stack.last().unwrap()
    }

    fn get_min(&self) -> i32 {
        *self.min_stack.last().unwrap()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn min_stack_case_1() {
        // arrange
        let expected1 = -3;
        let expected2 = 0;
        let expected3 = -2;
        let mut obj = MinStack::new();

        // act
        obj.push(-2);
        obj.push(0);
        obj.push(-3);
        let actual1 = obj.get_min();
        obj.pop();
        let actual2: i32 = obj.top();
        let actual3: i32 = obj.get_min();

        // assert
        assert_eq!(expected1, actual1);
        assert_eq!(expected2, actual2);
        assert_eq!(expected3, actual3);
    }

    #[test]
    fn min_stack_case_2() {
        // arrange
        let expected1 = -2;
        let expected2 = -1;
        let expected3 = -2;
        let mut obj = MinStack::new();

        // act
        obj.push(-2);
        obj.push(0);
        obj.push(-1);
        let actual1 = obj.get_min();
        let actual2: i32 = obj.top();
        obj.pop();
        let actual3: i32 = obj.get_min();

        // assert
        assert_eq!(expected1, actual1);
        assert_eq!(expected2, actual2);
        assert_eq!(expected3, actual3);
    }
}
